validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java

```diff
@@ -182,16 +205,43 @@ private void scheduleAggregation(
 
   @Override
   public void onBlockProductionDue(final UnsignedLong slot) {
-    scheduledDuties.produceBlock(slot);
+    whenDutiesScheduled(slot, scheduledDuties::produceBlock);
   }
 
   @Override
   public void onAttestationCreationDue(final UnsignedLong slot) {
-    scheduledDuties.produceAttestations(slot);
+    whenDutiesScheduled(slot, scheduledDuties::produceAttestations);
   }
 
   @Override
   public void onAttestationAggregationDue(final UnsignedLong slot) {
-    scheduledDuties.performAggregation(slot);
+    whenDutiesScheduled(slot, scheduledDuties::performAggregation);
+  }
+
+  private void whenDutiesScheduled(final UnsignedLong slot, final Consumer<UnsignedLong> action) {
+    // We chain the futures to ensure all actions always happen in their original order.
+    final UnsignedLong epoch = compute_epoch_at_slot(slot);
+    final SafeFuture<Void> delayedAction =
+        pendingTasksByEpoch.computeIfPresent(
+            epoch, (key, previousTask) -> previousTask.thenRun(() -> action.accept(slot)));
+    if (delayedAction == null) {
+      // There was no pending tasks so execute immediately.
+      action.accept(slot);
+    } else {
+      removeWhenAllTasksComplete(epoch, delayedAction);
+    }
+  }
+
+  /**
+   * Once each new task has been added to {@link #pendingTasksByEpoch}, this function ensures that
+   * the task queue is removed when that tasks completes, if and only if it is still the last task
+   * in the queue.
+   *
+   * @param epoch the epoch the task was queued for
+   * @param enqueuedTask the task that was queued
+   */
+  private void removeWhenAllTasksComplete(
+      final UnsignedLong epoch, final SafeFuture<Void> enqueuedTask) {
+    enqueuedTask.always(() -> pendingTasksByEpoch.remove(epoch, enqueuedTask));
```

Comment:
Isn't there any side effect of chaining multiple `.always()` to the same Future chain? So they don't get overwritten or anything?

Reply:
`.always` winds up mapping to a `.handle` call under the covers - it's just like any other kind of chaining call on `CompletableFuture`, so perfectly fine to call multiple times.  In this case though it's actually only called once per future anyway because it's a different future passed into this function each time.